crates/env/src/engine/e2e/xts.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright 2018-2022 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    log_info,
+    sr25519,
+    ContractExecResult,
+    ContractInstantiateResult,
+    IdentifyAccount,
+    Signer,
+    Verify,
+};
+use crate::Environment;
+
+use core::marker::PhantomData;
+use jsonrpsee::{
+    core::client::ClientT,
+    rpc_params,
+    ws_client::{
+        WsClient,
+        WsClientBuilder,
+    },
+};
+use sp_core::{
+    Bytes,
+    H256,
+};
+use subxt::{
+    tx::{
+        ExtrinsicParams,
+        TxEvents,
+    },
+    OnlineClient,
+};
+
+/// The gas limit for contract instantiate and call dry runs.
+const DRY_RUN_GAS_LIMIT: u64 = 500_000_000_000;
+
+// TODO(#xxx) Should be fetched automatically.
+#[subxt::subxt(runtime_metadata_path = "metadata/contracts-node.scale")]
+pub(super) mod api {}
+
+/// A raw call to `pallet-contracts`'s `instantiate_with_code`.
+#[derive(Debug, scale::Encode, scale::Decode)]
+pub struct InstantiateWithCode<B> {
+    #[codec(compact)]
+    value: B,
+    #[codec(compact)]
+    gas_limit: crate::types::Gas,
+    storage_deposit_limit: Option<B>,
+    code: Vec<u8>,
+    data: Vec<u8>,
+    salt: Vec<u8>,
+}
```

Comment:
Shouldn't you use the generated data from the subxt macro here?

Reply:
By generated data you mean. that the `InstantiateWithCode` struct should be available in some generated form from `subxt`?

Reply:
Yes. This is the whole point of reading in the metadata. Use `api::tx().contracts().call()` to create your `StaticTxPayload`.

Reply:
Yes this would be the ideal way to do it. However I believe you would need to access the non-public `call_data` field for this, since [`StaticTxPayload::encode_call_data`](https://github.com/paritytech/subxt/blob/5ff849318b75ba32ca905ec9d34cc1d16a858b31/subxt/src/tx/tx_payload.rs#L81) prepends the pallet and call indices.

I believe there is a `subxt` release coming shortly, so you should get this in quickly.

Reply:
After implementing it I ran into a problem that makes it currently not possible to use the from-metadata-generated `InstantiateWithCode` & co structs. The reason is that those structs have a well-defined type, e.g. `type Balance = u128`. 

My E2E methods on the other hand are generalized, they just require `value: <E as Environment>::Balance`. In order to make my testing methods work with the generated structs I would have to require `value: u128` hardcoded. I don't want to do this though, as it would limit the E2E testing framework to support only chains with `value: u128`. 

I've discussed it with Andrew and we agree that the best option is to leave as is.

I still need to generate the types via `#[subxt::subxt(runtime_metadata_path = "metadata/contracts-node.scale")]` though, as they are required to determine:
```rust
.as_event::<xts::api::system::events::ExtrinsicFailed>()
```